sae/blocks.go

```diff
@@ -5,42 +5,240 @@ package sae
 
 import (
 	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"time"
 
 	"github.com/ava-labs/avalanchego/database"
 	"github.com/ava-labs/avalanchego/ids"
 	"github.com/ava-labs/avalanchego/snow/engine/snowman/block"
+	"github.com/ava-labs/avalanchego/vms/components/gas"
+	"github.com/ava-labs/libevm/common"
+	"github.com/ava-labs/libevm/core"
+	"github.com/ava-labs/libevm/core/state"
+	"github.com/ava-labs/libevm/core/txpool"
 	"github.com/ava-labs/libevm/core/types"
 	"github.com/ava-labs/libevm/rlp"
+	"github.com/holiman/uint256"
+	"go.uber.org/zap"
 
 	"github.com/ava-labs/strevm/blocks"
+	"github.com/ava-labs/strevm/params"
+	"github.com/ava-labs/strevm/txgossip"
 )
 
-// ParseBlock parses the buffer as [rlp] encoding of a [types.Block].
-func (vm *VM) ParseBlock(context.Context, []byte) (*blocks.Block, error) {
-	ethB := new(types.Block)
-	_ = rlp.DecodeBytes(nil, ethB)
-	return nil, errUnimplemented
+func (vm *VM) newBlock(eth *types.Block, parent, lastSettled *blocks.Block) (*blocks.Block, error) {
+	return blocks.New(eth, parent, lastSettled, vm.log())
+}
+
+// ParseBlock parses the buffer as [rlp] encoding of a [types.Block]. It does
+// NOT populate the block ancestry, which is done by [VM.VerifyBlock] i.f.f.
+// verification passes.
+func (vm *VM) ParseBlock(ctx context.Context, buf []byte) (*blocks.Block, error) {
+	b := new(types.Block)
+	if err := rlp.DecodeBytes(buf, b); err != nil {
+		return nil, fmt.Errorf("rlp.DecodeBytes(..., %T): %v", b, err)
+	}
+	return vm.newBlock(b, nil, nil)
```

Comment:
This seems to break the documented invariant on `ancestry`:
> // Invariant: ancestry is non-nil and contains non-nil pointers i.f.f. the
	// block hasn't itself been settled. A synchronous block (e.g. SAE genesis
	// or the last pre-SAE block) is always considered settled.

This block hasn't been settled, so it must contain non-nil pointers.

Of course, this can't actually be guaranteed here... So I think we need to update these invariants (or make some bigger change).